pkg/util/kubernetes/apiserver/events.go

```diff
@@ -7,103 +7,144 @@
+			Limit:                eventCardinalityLimit,
+			IncludeUninitialized: false,
+			FieldSelector:        filter,
+		})
+		if err != nil {
+			log.Errorf("Error Listing events: %s", err.Error())
+			return nil, srv, st, err
+		}
+		for _, e := range evList.Items {
+			// List call returns a different type than the Watch call.
+			added = append(added, &e)
+		}
+		st = time.Now()
+		srv = evList.ResourceVersion
+		return added, srv, st, nil
 	}
-
-	log.Tracef("Starting watch of events with resourceVersion %s", since)
-
-	eventWatcher, err := c.Cl.CoreV1().Events(metav1.NamespaceAll).Watch(metav1.ListOptions{Watch: true, ResourceVersion: since})
+	// Start watcher with the most up to date RV
+	evWatcher, err := c.Cl.CoreV1().Events(metav1.NamespaceAll).Watch(metav1.ListOptions{
+		Watch:                true,
+		ResourceVersion:      srv,
+		Limit:                eventCardinalityLimit,
+		IncludeUninitialized: false,
+		FieldSelector:        filter,
+	})
 	if err != nil {
-		return nil, nil, "0", fmt.Errorf("Failed to watch events: %v", err)
+		return added, srv, st, err
 	}
-	defer eventWatcher.Stop()
 
-	// To account for slow starts, wait longer for the first event
-	watcherTimeout := time.NewTimer(eventReadTimeout * 2)
+	defer evWatcher.Stop()
+	log.Debugf("Starting to watch from %s", srv)
+	// watch during 2 * timeout maximum and store where we are at.
+	timeoutParse := time.NewTimer(time.Duration(eventReadTimeout*2) * time.Second)
 	for {
 		select {
-		case rcvdEv, ok := <-eventWatcher.ResultChan():
+		case rcv, ok := <-evWatcher.ResultChan():
 			if !ok {
-				log.Debugf("Unexpected watch close")
-				return added, modified, strconv.Itoa(resVersionInt), nil
+				return added, srv, st, fmt.Errorf("Unexpected watch close")
 			}
-			if rcvdEv.Type == watch.Error {
-				status, ok := rcvdEv.Object.(*metav1.Status)
+			if rcv.Type == watch.Error {
+				status, ok := rcv.Object.(*metav1.Status)
 				if !ok {
-					return nil, nil, "0", errors.New("could not parse status of error event from the API Server.") // TODO custom error
+					return added, srv, st, fmt.Errorf("Could not unmarshall the status of the event")
 				}
-				if status.Reason == "Expired" {
-					// Known issue with ETCD https://github.com/kubernetes/kubernetes/issues/45506
-					// Once we have started the event watcher, we do not expect the resversion to be "0"
-					// Except when initializing. Forcing to 0 so we can keep on watching without hitting the cache.
-					log.Tracef("Resversion expired: %s", status.Message)
-					return added, modified, "0", nil
+				switch status.Reason {
+				// Using a switch as there are a lot of different types and we might want to explore adapting the behaviour for certain ones in the future.
+				case "Expired":
+					log.Debugf("Resource Version is too old, listing all events and collecting only the new ones")
+					evList, err := c.Cl.CoreV1().Events(metav1.NamespaceAll).List(metav1.ListOptions{
```

Comment:
isn't it similar to when we're listing above? maybe it should be in a function

Reply:
I did not do a function as I use a diff after the second call, but I can factorize.